### PR TITLE
Shift Spine On Page Turn for More Fluid Movement

### PR DESCRIPTION
**Pull request by owntheweb**
I really like the flippbook example and want to use it for a fun project. Beforehand, I feel it would be good to contribute back a little and get the example updated. I'll keep picking at this draft PR for a bit if that's cool. Comments and suggestions are welcome.

To start, I revised so that the "spine" moves with the pages so that more pages can be added and positioned without odd turning effects (e.g. page lifted up, flipped, then dropped down).

I also moved more page flipping control to the Book class with less of that in the app.

There was also an instance where the child Page class was calling a method on the parent Book class. Instead I called a callback from the child class that the parent could listen for.

Sample images show more clearly with a filter update (could be even clearer with less compressed images).

Pardon the line break conversion. Whitespace can be ignored when viewing via the sprocket/options to make it look much better. I can work to convert back to old `\r\n` format if needed.

I'll keep playing with it, updating to ES6 soon?

### CODE REVIEW

File: examples/Three/flipbook3/FlipBook3D.js [hunk 1]
```diff
@@ -1,262 +1,297 @@
...
+            this.updateSpine();
+        },
+    
+        // keep stacked pages from overlapping in the same space on flip
+        updateSpine: function() {
+            const self = this;
+    
+            for (const page of self.pages) {
+                if (page.index > self.index) {
+                    page.position.z = -((page.zz * page.index) - ((page.zz * self.index - 1)))
```
Comment (ChrisAtBombBomb) [inline examples/Three/flipbook3/FlipBook3D.js:72]: I have an extra `()` here hmm.

File: examples/Three/flipbook3/FlipBook3D.js [hunk 2]
```diff
@@ -1,262 +1,297 @@
...
+        pageColor : null,
+        sides : null,
+        onFinishCallback: () => {},
+    
+        flipLeft: function(pt) {
+            var self = this;
+    
+            if (
+                !self.flippingLeft &&
+                !self.flippingRight
```
Comment (ChrisAtBombBomb) [inline examples/Three/flipbook3/FlipBook3D.js:203]: I may need to revise this again. When flipping back and forth rapidly mid-book, things get messed up.

File: examples/Three/flipbook3/FlipBook3D.js [hunk 3]
```diff
@@ -1,262 +1,297 @@
...
-    },
-
-    flipRight: function(pt) {
-        var self = this;
-
-        if (
-            !self.isFlippedRight &&
-            !self.flippingRight &&
-            !self.flippingLeft &&
-            (self.index === self.book.getNumPages()-self.book.flippedright-1)
```
Comment (ChrisAtBombBomb) [inline examples/Three/flipbook3/FlipBook3D.js:191]: same here

File: examples/Three/flipbook3/FlipBook3D.js [hunk 4]
```diff
@@ -1,262 +1,297 @@
...
+        flipFinished: function() {
+            var t = this, self = t.page;
+            if (self.flippingLeft)
             {
-                //this.flipPt=pt;//e.localY/this.book.pageHeight;
+                self.flippingLeft = false;
+                self.isFlippedLeft = true;
+                self.flippingRight = false;
+                self.isFlippedRight = false;
+                // self.position.z = -self.zz*(self.book.getNumPages()-self.index);
```
Comment (ChrisAtBombBomb) [inline examples/Three/flipbook3/FlipBook3D.js:276]: to remove

File: examples/Three/flipbook3/FlipBook3D.js [hunk 5]
```diff
@@ -1,262 +1,297 @@
...
-        {
-            if (null != pt)
+            else if (self.flippingRight)
             {
-                //this.flipPt=pt;//e.localY/this.book.pageHeight;
+                self.flippingLeft = false;
+                self.isFlippedRight = true;
+                self.flippingRight = false;
+                self.isFlippedLeft = false;
+                // self.position.z = -self.zz*self.index;
```
Comment (ChrisAtBombBomb) [inline examples/Three/flipbook3/FlipBook3D.js:284]: to remove

**Comment by foo123** [comment]
I tend for my projects and classes to be as backwards compatible as possible (yes, still). So I tend to avoid the more current features, like ES6. At least for now. So I would appreciate if you can maintain it, without using ES6 or features that may possibly break backwards comaptibility.

**Comment by ChrisAtBombBomb** [comment]
Alright @foo123 I feel like this is good to go. I made a few more optimizations and did just a touch more cleanup with spacing (also not crazy with the whitespace, left spacing the same). There was a touch of ES6 sneaking in that was removed. Let me know if you would prefer this all to be merged into one commit or leave free-flowing as-is. Thanks again for a great project!

**Comment by foo123** [comment]
Can you make it one single commit, let's not clutter the repositories with unnecessary commits. Thanks

**Comment by ChrisAtBombBomb** [comment]
All set for one commit. Thanks again

**Pull request merged**